File: Graph/GFG/Medium/KahnAlgorithm/KahnAlgorithm.java
// https://www.geeksforgeeks.org/problems/topological-sort/1

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;

public class KahnAlgorithm {
    static int[] topoSort(int V, ArrayList<ArrayList<Integer>> adj) 
    {
        // add your code here
        int[] inDegree = new int[V];
        for(int i=0; i<V; i++) {
            for(int neighbor: adj.get(i)) {
                inDegree[neighbor]++;
            }
        }
        Deque<Integer> q = new ArrayDeque<>();
        for(int i=0; i<V; i++) {
            if(inDegree[i] == 0) {
                q.offer(i);
            }
        }
        int ans[] = new int[V];
        int index = 0;
        while(!q.isEmpty()) {
            int curr = q.poll();
            ans[index++] = curr;
            for(int neighbor: adj.get(curr)) {
                inDegree[neighbor]--;
                if(inDegree[neighbor] == 0) {
                    q.offer(neighbor);
                }
            }
        }
        return ans;
    }
}

File: Graph/GFG/Medium/KahnAlgorithm/KahnAlgorithm.cpp
// https://www.geeksforgeeks.org/problems/topological-sort/1
#include <bits/stdc++.h>
using namespace std;
vector<int> topoSort(int V, vector<int> adj[])
{
    // code here
    vector<int> inDegree(V, 0);
    for (int i = 0; i < V; i++)
    {
        for (int neighbor : adj[i])
        {
            inDegree[neighbor]++;
        }
    }
    queue<int> q;
    for (int i = 0; i < V; i++)
    {
        if (!inDegree[i])
        {
            q.push(i);
        }
    }
    vector<int> ans;
    while (!q.empty())
    {
        int curr = q.front();
        q.pop();
        ans.push_back(curr);
        for (int neighbor : adj[curr])
        {
            inDegree[neighbor]--;
            if (!inDegree[neighbor])
            {
                q.push(neighbor);
            }
        }
    }
    return ans;
}
int main(void)
{

    return 0;
}